vunit/ui.py

```diff
@@ -461,6 +461,13 @@ def add_source_files_from_csv(self, project_csv_path, vhdl_standard=None):
                     LOGGER.error("More than one library and one file in csv description")
         return files
 
+    def set_objects(self, files):

```

Comment:
The name set_objects is to general. You assume the object file will contain a VHPI implementation of write_byte and read_byte.

Reply:
This is not required. It is a general purpose method to add pre-built objects to the linker args in the elaboration command. I can be used to wrap GHDL in a C program, to provide the implementation of VHPI functions/procedures or for both features at the same time. In this example, both features are used. But that's not a requirement.

Moreover, the provided VHPI functions can be any other defined by the user in the testbench or in custom packages. This is because VHPI functions declared in multiple different VHDL sources/modules can then be provided all together in a single fat object or in multiple slimmer objects.

Nonetheless, I agree that the name could be better. But I think that the name is tightly related to how does the user tell that VHPI is to be used. E.g., the user might want to use VHPI for custom functions, but not have the 'external' vectors/memories. In this context, the objects added through `set_objects` are not required to define neither `write_byte` nor `read_byte`. Precisely, `extmem_pkg-novhpi` should be added to the project.

Reply:
My comment was based on the observation that it set `self._project.vhpi = True` which in your code implied that there was an external byte array implementation. It was thus not a general solution. The user could not use a set_object call to add another VHPI-object unrelated to the byte array.

If we should have a general `set_object` method on the VUnit API I think it should be a separate PR since it is not strictly necessary to achieve the goal of this PR. Also we need to think about how it is presented to the user. It just boils down to a sim_option which the user could unintentionally overwrite with another set_sim_option call. It would not be clear to the user how set_object would conflict with set_sim_option.

Reply:
> My comment was based on the observation that it set `self._project.vhpi = True` which in your code implied that there was an external byte array implementation. It was thus not a general solution. The user could not use a set_object call to add another VHPI-object unrelated to the byte array.

You are correct, my wording was not good. I meant that it is the placeholder for a general solution. Not that the actual implementation is general as I want it to be.
 
> If we should have a general `set_object` method on the VUnit API I think it should be a separate PR since it is not strictly necessary to achieve the goal of this PR.

I do think it is necessary to achieve this goal. The mechanism to provide an object is the same, independently of which are the functions implemented in it. In this PR we need a mechanism to let the user provide an object, at least.

Nevertheless, it is ok to handle it in a separate PR and then rebase this on top of it.

> Also we need to think about how it is presented to the user. It just boils down to a sim_option which the user could unintentionally overwrite with another set_sim_option call. It would not be clear to the user how set_object would conflict with set_sim_option.

As said, the current implementation is far from being good. I want to trigger this discussion we are having.

For GHDL, the requirement is to provide a sim_option. As you point out, the current implmentation allows the user to unwantedly overwrite the sim_option. As an alternative, I was considering to implement `set_object` and `add_object` as methods that manipulate an attribute of type list (probably named `_objects`) to Configuration. Then, the list would be added in the `_get_command` method of GHDLInterface. So, even if it is a sim_option, it would be treated as a different set.

This is because I think that the integration with other simulators might require handling the objects differently than adding them as a sim_option. For example, with ModelSim and FLI it seems that objects need to be in the path of the binary, but no CLI arg is required: https://opencores.org/usercontent/doc/1380917197, https://gitlab.com/reds-public/fss/tree/master/fss_demo_uart.

Reply:
See #481.